packages/trie/src/types.ts

```diff
@@ -24,11 +25,18 @@ export interface TrieOpts {
   /**
    * A database instance.
    */
-  db?: DB
+  db?: DB | CheckpointDB
+
+  /**
+   * A database instance.
+   */
+  dbStorage?: DB
```

Comment:
Ugh.

This makes things highly complicated and I do not think that people "get" why we (they) need two database options here and where their trie is ultimately safed (to be honest: I am not getting both questions either right now, might "get it" when further looking at the code.

So - again writing this without having the internal picture: what I would want here is that there is just *one* database to be passed in and everything else is handled internally and shielded from the user - if e.g. a temporary DB for state changes or whatever needs to be created.

Some side context info: I found this already not so lucky in the original CheckpointTrie code base, so this separate naming for the storage DB with `dbStorage` - I can remember that I stumbled upon this along original PR review - but I could live with that as long as it stayed internal.

Reply:
Ah.

I had to look at this a couple of times but I think I am slowly getting confidence. 🙂 

Yes, that seems to be at least going in a very good direction - first couldn't believe it - but maybe that's already there to a greater extend.

I would be inclined to also not allow passing in a `CheckpointDB` directly for `db` in the constructor (at least for now, we could add this later if we want). This would make it slicker and more clear and we do not have to explain to people how to create their own CheckpointDBs. (or is there a concrete use case (e.g. also by you) for this *right now*?).

For now always using a CheckpointDB: so is it double/triple-checked that a CP DB (so in particular the `db/checkpoint.ts` class) behaves exactly as a non-CP-DB in an non-checkpointed context? 🤔 Should very muhc be since you could use this without CPing before, but better triple-ask here to get really really sure.

It is just super-essential that people do not get their DBs corrupted.